File: app/src/main/java/ch/bailu/aat/util/DownloadApi.java
package ch.bailu.aat.util;

import android.content.Context;

import java.io.UnsupportedEncodingException;

import ch.bailu.aat.services.InsideContext;
import ch.bailu.aat.services.ServiceContext;
import ch.bailu.aat.services.background.BackgroundService;
import ch.bailu.aat.services.background.BackgroundTask;
import ch.bailu.aat.services.background.DownloadTask;
import ch.bailu.aat.util.ui.AppLog;
import ch.bailu.util_java.foc.Foc;

public abstract class DownloadApi extends OsmApiHelper {

    private BackgroundTask task = BackgroundTask.NULL;

    private static class ApiQueryTask extends DownloadTask {
        private final String queryString;
        private final Foc queryFile;

        public ApiQueryTask(Context c, String source, Foc target, String qs, Foc qf) {
            super(c, source, target);
            queryString = qs;
            queryFile   = qf;
        }


        @Override
        public long bgOnProcess(ServiceContext sc) {
            try {
                long size = bgDownload();
                TextBackup.write(queryFile, queryString);

                AppBroadcaster.broadcast(sc.getContext(),
                        AppBroadcaster.FILE_CHANGED_ONDISK, getFile(), getSource());

                return size;
            } catch (Exception e) {
                logError(e);
                return 1;
            }
        }

        @Override
        protected void logError(Exception e) {
            AppLog.e(getContext(), e);
        }
    }





    @Override
    public void startTask(ServiceContext scontext) {
        new InsideContext(scontext) {
            @Override
            public void run() {

                try {
                    BackgroundService background = scontext.getBackgroundService();

                    final String query = getQueryString();
                    task = new ApiQueryTask(
                            scontext.getContext(),
                            getUrl(query),
                            getResultFile(),
                            query,
                            getQueryFile());
                    background.process(task);
                } catch (UnsupportedEncodingException e) {
                    e.printStackTrace();
                }
            }
        };
    }

    protected abstract String getQueryString();


    @Override
    public Exception getException() {
        return task.getException();
    }

}

File: aat-lib/src/main/java/ch/bailu/aat_lib/api/DownloadApi.kt
package ch.bailu.aat_lib.api

import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.coordinates.BoundingBoxE6
import ch.bailu.aat_lib.logger.AppLog
import ch.bailu.aat_lib.service.background.BackgroundTask
import ch.bailu.aat_lib.service.background.DownloadTask
import ch.bailu.aat_lib.util.fs.TextBackup
import ch.bailu.foc.Foc
import java.io.UnsupportedEncodingException

abstract class DownloadApi : OsmApiConfiguration() {
    private var task = BackgroundTask.Companion.NULL

    private class ApiQueryTask(c: AppContext, source: String, target: Foc, private val queryString: String, private val queryFile: Foc) : DownloadTask(source, target, c.downloadConfig) {

        override fun bgOnProcess(appContext: AppContext): Long {
            return try {
                val size = bgDownload()
                TextBackup.Companion.write(queryFile, queryString)
                appContext.broadcaster.broadcast(
                    AppBroadcaster.FILE_CHANGED_ONDISK, getFile().toString(), source.toString()
                )
                size
            } catch (e: Exception) {
                AppLog.e(this, e) // TODO user friendly message
                1
            }
        }
    }

    override fun startTask(appContext: AppContext, boundingBoxE6: BoundingBoxE6) {
        appContext.services.insideContext {
            try {
                val background = appContext.services.getBackgroundService()
                val query: String = queryString
                val url = getUrl(query, boundingBoxE6)
                task = ApiQueryTask(
                    appContext,
                    url,
                    resultFile,
                    query,
                    queryFile
                )
                background.process(task)
            } catch (e: UnsupportedEncodingException) {
                AppLog.e(this, e) // TODO user friendly message
            }
        }
    }

    protected abstract val queryString: String
}
